File: src/fwk/toolkit/metadatawidget.cpp
/*
 * niepce - fwk/toolkit/metadatawidget.cpp
 *
 * Copyright (C) 2008-2022 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <utility>

#include <boost/lexical_cast.hpp>
#include <boost/format.hpp>
#include <boost/rational.hpp>

#include <glibmm/i18n.h>
#include <gtkmm/entry.h>
#include <gtkmm/eventcontrollerfocus.h>
#include <gtkmm/label.h>
#include <gtkmm/textview.h>

#include "fwk/base/debug.hpp"
#include "fwk/base/autoflag.hpp"
#include "fwk/utils/exempi.hpp"
#include "fwk/utils/stringutils.hpp"
#include "fwk/toolkit/widgets/notabtextview.hpp"
#include "fwk/toolkit/widgets/tokentextview.hpp"

#include "rust_bindings.hpp"

#include "metadatawidget.hpp"

namespace fwk {

MetaDataWidget::MetaDataWidget(const Glib::ustring & title)
    : ToolboxItemWidget(title),
      m_fmt(nullptr),
      m_update(false)
{
    m_table.set_column_homogeneous(true);
    m_table.set_row_homogeneous(false);
    m_table.insert_column(0);
    m_table.insert_column(0);
    m_table.set_margin_start(8);
    set_child(m_table);
    set_sensitive(false);
}

void MetaDataWidget::set_data_format(const MetaDataSectionFormat* fmt)
{
    m_fmt = fmt;
    create_widgets_for_format(fmt);
}

void MetaDataWidget::rating_callback(GtkWidget* w, gint rating, gpointer user_data)
{
    auto self = static_cast<MetaDataWidget*>(user_data);
    auto id = GPOINTER_TO_INT(g_object_get_data(G_OBJECT(w), "id"));
    self->on_int_changed(rating, static_cast<ffi::NiepcePropertyIdx>(id));
}

Gtk::Widget*
MetaDataWidget::create_star_rating_widget(bool readonly, ffi::NiepcePropertyIdx id)
{
    Gtk::Widget* r =
        Gtk::manage(Glib::wrap(
                        GTK_WIDGET(ffi::fwk_rating_label_new(0, !readonly))));
    if (!readonly) {
        r->set_data("id", GINT_TO_POINTER(id));
        g_signal_connect(r->gobj(), "rating-changed", G_CALLBACK(rating_callback), this);
    }
    return r;
}

Gtk::Widget*
MetaDataWidget::create_string_array_widget(bool readonly, ffi::NiepcePropertyIdx id)
{
    fwk::TokenTextView* ttv = Gtk::manage(new fwk::TokenTextView());
    if (!readonly) {
        auto ctrl = Gtk::EventControllerFocus::create();
        ctrl->signal_leave().connect([this, ttv, id] {
            this->on_string_array_changed(ttv, id);
        });
        ttv->add_controller(ctrl);
    }
    return ttv;
}

Gtk::Widget*
MetaDataWidget::create_text_widget(bool readonly, ffi::NiepcePropertyIdx id)
{
    if(readonly) {
        Gtk::Label * l = Gtk::manage(new Gtk::Label());
        l->set_xalign(0.0f);
        l->set_yalign(0.5f);
        // This will prevent the label from being expanded.
        l->set_ellipsize(Pango::EllipsizeMode::MIDDLE);
        return l;
    }

    Gtk::TextView * e = Gtk::manage(new NoTabTextView());
    e->set_editable(true);
    e->set_wrap_mode(Gtk::WrapMode::WORD);
    auto ctrl = Gtk::EventControllerFocus::create();
    e->add_controller(ctrl);
    auto buffer = e->get_buffer();
    ctrl->signal_leave().connect([this, buffer, id] {
        this->on_text_changed(buffer, id);
    });
    return e;
}

Gtk::Widget*
MetaDataWidget::create_string_widget(bool readonly, ffi::NiepcePropertyIdx id)
{
    if(readonly) {
        Gtk::Label * l = Gtk::manage(new Gtk::Label());
        l->set_xalign(0.0f);
        l->set_yalign(0.5f);
        // This will prevent the label from being expanded.
        l->set_ellipsize(Pango::EllipsizeMode::MIDDLE);
        return l;
    }

    Gtk::Entry * e = Gtk::manage(new Gtk::Entry());
    e->set_has_frame(false); // TODO make that a custom widget
    auto ctrl = Gtk::EventControllerFocus::create();
    e->add_controller(ctrl);
    ctrl->signal_leave().connect([this, e, id] {
        this->on_str_changed(e, id);
    });
    return e;
}

Gtk::Widget*
MetaDataWidget::create_date_widget(bool /*readonly*/, ffi::NiepcePropertyIdx id)
{
    // for now a date widget is just like a string. Read only
    return create_string_widget(true, id);
}

void
MetaDataWidget::create_widgets_for_format(const MetaDataSectionFormat * fmt)
{
    Gtk::Widget *w = nullptr;
    auto current = fmt->formats.begin();
    int n_row = 0;

    while(current != fmt->formats.end()) {
        Gtk::Label *labelw = Gtk::manage(new Gtk::Label(
                                             Glib::ustring("<b>")
                                             + current->label + "</b>"));
        if(current->type != MetaDT::STRING_ARRAY
           && current->type != MetaDT::TEXT) {
            labelw->set_xalign(0.0f);
            labelw->set_yalign(0.5f);
        }
        else {
            // Text can wrap. Different alignment for the label
            labelw->set_xalign(0.0f);
            labelw->set_yalign(0.0f);
        }
        labelw->set_use_markup(true);

        switch(current->type) {
        case MetaDT::STAR_RATING:
            w = create_star_rating_widget(current->readonly, current->id);
            break;
        case MetaDT::STRING_ARRAY:
            w = create_string_array_widget(current->readonly, current->id);
            break;
        case MetaDT::TEXT:
            w = create_text_widget(current->readonly, current->id);
            break;
        case MetaDT::DATE:
            w = create_date_widget(current->readonly, current->id);
            break;
        default:
            w = create_string_widget(current->readonly, current->id);
            break;
        }

        m_table.insert_row(n_row + 1);
        m_table.attach(*labelw, 0, n_row, 1, 1);
        m_table.attach_next_to(*w, *labelw, Gtk::PositionType::RIGHT, 1, 1);
        m_data_map.insert(std::make_pair(current->id, w));

        current++;
        n_row++;
    }
}

void MetaDataWidget::clear_widget(const std::pair<const ffi::NiepcePropertyIdx, Gtk::Widget *> & p)
{
    AutoFlag flag(m_update);
    Gtk::Label * l = dynamic_cast<Gtk::Label*>(p.second);
    if(l) {
        l->set_text("");
        return;
    }
    Gtk::Entry * e = dynamic_cast<Gtk::Entry*>(p.second);
    if(e) {
        e->set_text("");
        return;
    }
    fwk::TokenTextView *ttv = dynamic_cast<fwk::TokenTextView*>(p.second);
    if(ttv) {
        ttv->set_tokens(fwk::TokenTextView::Tokens());
        return;
    }
    Gtk::TextView * tv = dynamic_cast<Gtk::TextView*>(p.second);
    if(tv) {
        tv->get_buffer()->set_text("");
        return;
    }
    Gtk::Widget* rl = dynamic_cast<Gtk::Widget*>(p.second);
    if (rl) {
        ffi::fwk_rating_label_set_rating(rl->gobj(), 0);
        return;
    }
}

void MetaDataWidget::set_data_source(const fwk::PropertyBagPtr& properties)
{
    DBG_OUT("set data source");
    m_current_data = properties;
    if(!m_data_map.empty()) {
        std::for_each(m_data_map.cbegin(), m_data_map.cend(),
                      [this] (const decltype(m_data_map)::value_type& p) {
                          this->clear_widget(p);
                      });
    }
    bool is_empty =
        static_cast<bool>(properties) ? eng_property_bag_is_empty(properties.get()) : true;
    set_sensitive(!is_empty);
    if(is_empty) {
        return;
    }
    if(!m_fmt) {
        DBG_OUT("empty format");
        return;
    }

    auto current = m_fmt->formats.begin();
    while (current != m_fmt->formats.end()) {
        auto result = get_value_for_property(*properties, current->id);
        if (!result.empty() || !current->readonly) {
            add_data(*current, std::move(result));
        }
        else {
            DBG_OUT("get_property failed id = %d, label = %s",
                    static_cast<uint32_t>(current->id), current->label);
        }
        current++;
    }
}

bool MetaDataWidget::set_fraction_dec_data(Gtk::Widget* w,
                                           const PropertyValuePtr& value)
{
    if (!fwk_property_value_is_string(value.get())) {
        ERR_OUT("Data not string(fraction)");
        return false;
    }
    try {
        const std::string str_value = fwk::property_value_get_string(*value);
        DBG_OUT("set fraction dec %s", str_value.c_str());
        std::string frac = str(boost::format("%.1f")
                               % fwk::fraction_to_decimal(str_value));
        AutoFlag flag(m_update);
        static_cast<Gtk::Label*>(w)->set_text(frac);
    }
    catch(...) {
        return false;
    }
    return true;
}

bool MetaDataWidget::set_fraction_data(Gtk::Widget* w,
                                       const PropertyValuePtr& value)
{
    if (!fwk_property_value_is_string(value.get())) {
        ERR_OUT("Data not string(fraction)");
        return false;
    }
    try {
        const std::string str_value = fwk::property_value_get_string(*value);
        DBG_OUT("set fraction %s", str_value.c_str());
        boost::rational<int> r
            = boost::lexical_cast<boost::rational<int>>(str_value);

        std::string frac = str(boost::format("%1%/%2%")
                               % r.numerator() % r.denominator());
        AutoFlag flag(m_update);
        static_cast<Gtk::Label*>(w)->set_text(frac);
    }
    catch(...) {
        return false;
    }
    return true;
}

bool MetaDataWidget::set_star_rating_data(Gtk::Widget* w,
                                          const PropertyValuePtr& value)
{
    if (!fwk_property_value_is_integer(value.get())) {
        ERR_OUT("Data not integer");
        return false;
    }
    try {
        int rating = fwk_property_value_get_integer(value.get());
        AutoFlag flag(m_update);
        ffi::fwk_rating_label_set_rating(static_cast<Gtk::Widget*>(w)->gobj(), rating);
    }
    catch(...) {
        return false;
    }
    return true;
}

bool MetaDataWidget::set_string_array_data(Gtk::Widget* w, bool readonly,
                                           const PropertyValuePtr& value)
{
    try {
        AutoFlag flag(m_update);
        std::vector<std::string> tokens = fwk::property_value_get_string_array(*value);

        static_cast<fwk::TokenTextView*>(w)->set_tokens(tokens);
        static_cast<fwk::TokenTextView*>(w)->set_editable(!readonly);
    }
    catch(...) {
        return false;
    }
    return true;
}

bool MetaDataWidget::set_text_data(Gtk::Widget* w, bool readonly,
                                   const PropertyValuePtr& value)
{
    if (!fwk_property_value_is_string(value.get())) {
        ERR_OUT("Data not string.");
        return false;
    }
    try {
        AutoFlag flag(m_update);
        if(readonly) {
            static_cast<Gtk::Label*>(w)->set_text(
                fwk::property_value_get_string(*value));
        } else {
            static_cast<Gtk::TextView*>(w)->get_buffer()->set_text(
                fwk::property_value_get_string(*value));
        }
    }
    catch(...) {
        return false;
    }
    return true;
}

bool MetaDataWidget::set_string_data(Gtk::Widget* w, bool readonly,
                                     const PropertyValuePtr& value)
{
    if (!fwk_property_value_is_string(value.get())) {
        ERR_OUT("Data not string.");
        return false;
    }
    try {
        AutoFlag flag(m_update);
        if(readonly) {
            static_cast<Gtk::Label*>(w)->set_text(
                fwk::property_value_get_string(*value));
        } else {
            static_cast<Gtk::Entry*>(w)->set_text(
                fwk::property_value_get_string(*value));
        }
    }
    catch(...) {
        return false;
    }
    return true;
}

bool MetaDataWidget::set_date_data(Gtk::Widget* w, const PropertyValuePtr& value)
{
    if (!fwk_property_value_is_date(value.get())) {
        return false;
    }
    try {
        AutoFlag flag(m_update);
        const fwk::Date* date = fwk_property_value_get_date(value.get());
        if (date) {
            static_cast<Gtk::Label*>(w)->set_text(fwk::date_to_string(date));

            DBG_OUT("setting date data %s", fwk::date_to_string(date).c_str());
        } else {
            return false;
        }
    }
    catch(...) {
        return false;
    }
    return true;
}

void MetaDataWidget::add_data(const MetaDataFormat& current,
                              fwk::Option<PropertyValuePtr>&& optional_value)
{
    if (optional_value.empty()) {
        return;
    }
    auto value = optional_value.unwrap();
    if (fwk_property_value_is_empty(value.get())) {
        return;
    }

    Gtk::Widget *w = nullptr;
    auto iter = m_data_map.find(current.id);
    if(iter == m_data_map.end()) {
        ERR_OUT("no widget for property");
        return;
    }

    w = static_cast<Gtk::Label*>(iter->second);

    switch(current.type) {
    case MetaDT::FRAC_DEC:
        set_fraction_dec_data(w, value);
        break;
    case MetaDT::FRAC:
        set_fraction_data(w, value);
        break;
    case MetaDT::STAR_RATING:
        set_star_rating_data(w, value);
        break;
    case MetaDT::STRING_ARRAY:
        set_string_array_data(w, current.readonly, value);
        break;
    case MetaDT::TEXT:
        set_text_data(w, current.readonly, value);
        break;
    case MetaDT::DATE:
        set_date_data(w, value);
        break;
    default:
        if (!set_string_data(w, current.readonly, value)) {
            ERR_OUT("failed to set value for %u", static_cast<uint32_t>(current.id));
        }
        break;
    }
}

bool MetaDataWidget::on_str_changed(Gtk::Entry *e,
                                    ffi::NiepcePropertyIdx prop)
{
    if(m_update) {
        return true;
    }
    emit_metadata_changed(prop, fwk::property_value_new(e->get_text()));
    return true;
}

bool MetaDataWidget::on_text_changed(Glib::RefPtr<Gtk::TextBuffer> b,
                                     ffi::NiepcePropertyIdx prop)
{
    if(m_update) {
        return true;
    }
    emit_metadata_changed(prop,
                          fwk::property_value_new(b->get_text()));
    return true;
}

bool MetaDataWidget::on_string_array_changed(fwk::TokenTextView * ttv,
                                             ffi::NiepcePropertyIdx prop)
{
    if(m_update) {
        return true;
    }
    fwk::TokenTextView::Tokens tok;
    ttv->get_tokens(tok);
    emit_metadata_changed(prop,
                          fwk::property_value_new(tok));
    return true;
}

void MetaDataWidget::on_int_changed(int value, ffi::NiepcePropertyIdx prop)
{
    if(m_update) {
        return;
    }
    emit_metadata_changed(prop, fwk::property_value_new(value));
}

void MetaDataWidget::emit_metadata_changed(ffi::NiepcePropertyIdx prop,
                                           const fwk::PropertyValuePtr & value)
{
    fwk::PropertyBagPtr props = fwk::property_bag_new();
    fwk::PropertyBagPtr old_props = fwk::property_bag_new();
    fwk::set_value_for_property(*props, prop, *value);
    auto result = fwk::get_value_for_property(*m_current_data, prop);
    if (!result.empty()) {
        fwk::set_value_for_property(*old_props, prop, *result.unwrap());
    }
    signal_metadata_changed.emit(props, old_props);
}

}

/*
  Local Variables:
  mode:c++
  c-file-style:"stroustrup"
  c-file-offsets:((innamespace . 0))
  indent-tabs-mode:nil
  fill-column:80
  End:
*/

File: crates/npc-fwk/src/toolkit/widgets/metadata_widget.rs
/*
 * niepce - fwk/toolkit/widgets/metadata_widget.rs
 *
 * Copyright (C) 2022-2025 Hubert Figuière
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use crate::glib;
use crate::gtk4;
use glib::prelude::*;
use gtk4::subclass::prelude::*;

use super::ToolboxItem;
use crate::PropertyBag;

pub type MetadataPropertyBag = PropertyBag<u32>;

/// A wrapped `MetadataPropertyBag` for use with `glib::Value` in signals.
#[derive(Clone, Default, glib::Boxed)]
#[boxed_type(name = "PropertyBag")]
pub struct WrappedPropertyBag(pub MetadataPropertyBag);

// This bridge content should be moved when the bridge is removed.
#[repr(u32)]
#[derive(Clone, PartialEq)]
pub enum MetaDT {
    #[allow(dead_code)]
    NONE = 0,
    #[allow(dead_code)]
    STRING,
    StringArray,
    TEXT,
    DATE,
    FRAC,
    FracDec, // Fraction as decimal
    StarRating,
    #[allow(dead_code)]
    SIZE, // Size in bytes
}

#[derive(Clone)]
pub struct MetadataFormat {
    pub label: String,
    pub id: u32, // NiepcePropertyIdx
    pub type_: MetaDT,
    pub readonly: bool,
}

#[derive(Clone)]
pub struct MetadataSectionFormat {
    pub section: String,
    pub formats: Vec<MetadataFormat>,
}

glib::wrapper! {
    pub struct MetadataWidget(
    ObjectSubclass<imp::MetadataWidget>)
    @extends ToolboxItem, gtk4::Box, gtk4::Widget,
    @implements gtk4::Accessible, gtk4::Buildable, gtk4::ConstraintTarget;
}

impl MetadataWidget {
    pub fn new(title: &str) -> MetadataWidget {
        let obj: MetadataWidget = glib::Object::new();
        obj.upcast_ref::<ToolboxItem>().set_title(title);

        obj
    }

    /// Connect to the signal `metadata-changed`
    pub fn connect_metadata_changed<F>(&self, f: F) -> glib::SignalHandlerId
    where
        F: Fn(&Self, WrappedPropertyBag, WrappedPropertyBag) + 'static,
    {
        self.connect_closure(
            "metadata-changed",
            true,
            glib::closure_local!(move |w, new, old| {
                f(&w, new, old);
            }),
        )
    }

    /// Set the data source of the metadata.
    pub fn set_data_source(&self, properties: Option<MetadataPropertyBag>) {
        self.imp().set_data_source(properties);
    }

    pub fn set_data_format(&self, fmt: Option<MetadataSectionFormat>) {
        self.imp().set_data_format(fmt);
    }
}

mod imp {
    use std::cell::RefCell;
    use std::collections::HashMap;

    use crate::glib;
    use crate::gtk4;
    use glib::subclass::*;
    use gtk4::prelude::*;
    use gtk4::subclass::prelude::*;

    use crate::PropertyValue;

    use super::super::prelude::*;
    use super::super::{RatingLabel, TokenTextView};
    use super::{
        MetaDT, MetadataFormat, MetadataPropertyBag, MetadataSectionFormat, WrappedPropertyBag,
    };

    fn clear_widget(widget: &gtk4::Widget) {
        if let Some(label) = widget.downcast_ref::<gtk4::Label>() {
            label.set_text("");
        } else if let Some(entry) = widget.downcast_ref::<gtk4::Entry>() {
            entry.set_text("");
        } else if let Some(ttv) = widget.downcast_ref::<TokenTextView>() {
            ttv.set_tokens(&[]);
        } else if let Some(tv) = widget.downcast_ref::<gtk4::TextView>() {
            tv.buffer().set_text("");
        } else if let Some(rating) = widget.downcast_ref::<RatingLabel>() {
            rating.set_rating(0);
        } else {
            err_out!("Unknow widget type {}", widget.type_().name());
        }
    }

    pub struct MetadataWidget {
        widget: gtk4::Grid,
        data_map: RefCell<HashMap<u32, gtk4::Widget>>,
        current_data: RefCell<Option<MetadataPropertyBag>>,
        fmt: RefCell<Option<MetadataSectionFormat>>,
    }

    impl MetadataWidget {
        fn create_star_rating_widget(&self, readonly: bool, id: u32) -> gtk4::Widget {
            let rating = RatingLabel::new(0, !readonly);
            if !readonly {
                let obj = self.obj();
                rating.connect_rating_changed(glib::clone!(
                    #[weak]
                    obj,
                    move |_, rating| {
                        obj.imp()
                            .emit_metadata_changed(id, &PropertyValue::Int(rating));
                    }
                ));
            }

            rating.upcast()
        }

        fn create_text_widget(&self, readonly: bool, id: u32) -> gtk4::Widget {
            if readonly {
                self.create_string_widget(readonly, id)
            } else {
                let entry = gtk4::TextView::new();
                entry.set_accepts_tab(false);
                entry.set_editable(true);
                entry.set_wrap_mode(gtk4::WrapMode::Word);
                let ctrl = gtk4::EventControllerFocus::new();
                let obj = self.obj();
                ctrl.connect_leave(glib::clone!(
                    #[weak]
                    entry,
                    #[weak]
                    obj,
                    move |_| {
                        let buffer = entry.buffer();
                        let start = buffer.start_iter();
                        let end = buffer.end_iter();
                        obj.imp().emit_metadata_changed(
                            id,
                            &PropertyValue::String(buffer.text(&start, &end, true).to_string()),
                        );
                    }
                ));

                entry.add_controller(ctrl);
                entry.upcast()
            }
        }
        fn create_string_widget(&self, readonly: bool, id: u32) -> gtk4::Widget {
            if readonly {
                let label = gtk4::Label::new(None);
                label.set_xalign(0.0);
                label.set_yalign(0.5);
                label.set_ellipsize(gtk4::pango::EllipsizeMode::Middle);

                label.upcast()
            } else {
                let entry = gtk4::Entry::new();
                entry.set_has_frame(false);
                let ctrl = gtk4::EventControllerFocus::new();

                let obj = self.obj();
                ctrl.connect_leave(glib::clone!(
                    #[weak]
                    entry,
                    #[weak]
                    obj,
                    move |_| {
                        obj.imp().emit_metadata_changed(
                            id,
                            &PropertyValue::String(entry.text().to_string()),
                        );
                    }
                ));

                entry.add_controller(ctrl);
                entry.upcast()
            }
        }

        fn create_string_array_widget(&self, readonly: bool, id: u32) -> gtk4::Widget {
            let ttv = TokenTextView::new();
            if !readonly {
                let ctrl = gtk4::EventControllerFocus::new();

                let obj = self.obj();
                ctrl.connect_leave(glib::clone!(
                    #[weak]
                    ttv,
                    #[weak]
                    obj,
                    move |_| {
                        obj.imp()
                            .emit_metadata_changed(id, &PropertyValue::StringArray(ttv.tokens()));
                    }
                ));
                ttv.add_controller(ctrl);
            }

            ttv.upcast()
        }

        fn create_date_widget(&self, readonly: bool, id: u32) -> gtk4::Widget {
            self.create_string_widget(readonly, id)
        }

        fn create_widgets_for_format(&self, fmt: &MetadataSectionFormat) {
            for (i, f) in fmt.formats.iter().enumerate() {
                let label = gtk4::Label::new(Some(&format!("<b>{}</b>", &f.label)));
                label.set_use_markup(true);
                label.set_xalign(0.0);
                if f.type_ != MetaDT::StringArray {
                    label.set_yalign(0.5);
                } else {
                    label.set_yalign(0.0);
                }
                let w = match f.type_ {
                    MetaDT::StarRating => self.create_star_rating_widget(f.readonly, f.id),
                    MetaDT::StringArray => self.create_string_array_widget(f.readonly, f.id),
                    MetaDT::TEXT => self.create_text_widget(f.readonly, f.id),
                    MetaDT::DATE => self.create_date_widget(f.readonly, f.id),
                    _ => self.create_string_widget(f.readonly, f.id),
                };
                let row = i as i32;
                self.widget.insert_row(row + 1);
                self.widget.attach(&label, 0, row, 1, 1);
                self.widget
                    .attach_next_to(&w, Some(&label), gtk4::PositionType::Right, 1, 1);
                self.data_map.borrow_mut().insert(f.id, w);
            }
        }

        fn add_data(&self, fmt: &MetadataFormat, value: &PropertyValue) {
            let data_map = self.data_map.borrow();
            let w = data_map.get(&fmt.id);

            if w.is_none() {
                err_out!("No widget for property {}", fmt.id);
                return;
            }

            let w = w.as_ref().unwrap();
            match fmt.type_ {
                MetaDT::FracDec => self.set_fraction_dec_data(w, value),
                MetaDT::FRAC => self.set_fraction_data(w, value),
                MetaDT::StarRating => self.set_star_rating_data(w, value),
                MetaDT::StringArray => self.set_string_array_data(w, fmt.readonly, value),
                MetaDT::TEXT => self.set_text_data(w, fmt.readonly, value),
                MetaDT::DATE => self.set_date_data(w, value),
                _ => {
                    if !self.set_text_data(w, fmt.readonly, value) {
                        err_out!("failed to set value for {}", fmt.id);
                        false
                    } else {
                        true
                    }
                }
            };
        }

        pub(super) fn set_data_format(&self, fmt: Option<MetadataSectionFormat>) {
            self.fmt.replace(fmt);
            if let Some(fmt) = self.fmt.borrow().as_ref() {
                self.create_widgets_for_format(fmt);
            }
            // XXX what if None? Should we delete the widgets?
        }

        pub(super) fn set_data_source(&self, properties: Option<MetadataPropertyBag>) {
            self.current_data.replace(properties);
            self.data_map.borrow().values().for_each(clear_widget);

            let is_empty = self
                .current_data
                .borrow()
                .as_ref()
                .map(|v| v.is_empty())
                .unwrap_or(true);
            self.obj().set_sensitive(!is_empty);
            if is_empty {
                return;
            }
            let properties = self.current_data.borrow();
            if let Some(fmt) = self.fmt.borrow().as_ref() {
                fmt.formats.iter().for_each(|f| {
                    if let Some(value) = properties.as_ref().and_then(|v| v.get(&f.id)) {
                        self.add_data(f, value)
                    } else {
                        // XXX value is empty
                    }
                });
            }
        }

        fn set_fraction_dec_data(&self, w: &gtk4::Widget, value: &PropertyValue) -> bool {
            if let Some(s) = value.string() {
                dbg_out!("set faction dec {}", s);
                return if let Some(w) = w.downcast_ref::<gtk4::Label>() {
                    let dec_str = crate::fraction_to_decimal(s)
                        .map(|dec| dec.to_string())
                        .unwrap_or_else(|| "NaN".to_string());

                    w.set_text(&dec_str);
                    true
                } else {
                    err_out!(
                        "Incorrect widget type for fraction_dec: {}",
                        w.type_().name()
                    );
                    false
                };
            }

            err_out!("Data not a string");
            false
        }

        fn set_fraction_data(&self, w: &gtk4::Widget, value: &PropertyValue) -> bool {
            if let Some(s) = value.string() {
                dbg_out!("set fraction {}", s);
                return if let Some(w) = w.downcast_ref::<gtk4::Label>() {
                    if let Some((n, d)) = crate::parse_fraction(s) {
                        let frac_str = format!("{n}/{d}");
                        w.set_text(&frac_str);
                        true
                    } else {
                        err_out!("Invalid fraction {}", s);
                        false
                    }
                } else {
                    err_out!("Incorrect widget type for fraction: {}", w.type_().name());
                    false
                };
            }

            err_out!("Data not a string");
            false
        }

        fn set_star_rating_data(&self, w: &gtk4::Widget, value: &PropertyValue) -> bool {
            if let Some(i) = value.integer() {
                return if let Some(w) = w.downcast_ref::<RatingLabel>() {
                    w.set_rating(i);
                    true
                } else {
                    err_out!("Incorrect widget type for rating: {}", w.type_().name());
                    false
                };
            }

            err_out!("Data not integer");
            false
        }

        fn set_string_array_data(
            &self,
            w: &gtk4::Widget,
            readonly: bool,
            value: &PropertyValue,
        ) -> bool {
            if let Some(tokens) = value.string_array() {
                return if let Some(w) = w.downcast_ref::<TokenTextView>() {
                    w.set_tokens(tokens);
                    w.set_editable(!readonly);
                    true
                } else {
                    err_out!(
                        "Incorrect widget type for string array: {}",
                        w.type_().name()
                    );
                    false
                };
            }
            err_out!("Data not string array");
            false
        }

        fn set_text_data(&self, w: &gtk4::Widget, readonly: bool, value: &PropertyValue) -> bool {
            if let Some(s) = value.string() {
                if readonly {
                    return if let Some(w) = w.downcast_ref::<gtk4::Label>() {
                        w.set_text(s);
                        true
                    } else {
                        err_out!("Incorrect widget type {}", w.type_().name());
                        false
                    };
                } else {
                    return if let Some(w) = w.downcast_ref::<gtk4::Entry>() {
                        w.set_text(s);
                        true
                    } else if let Some(w) = w.downcast_ref::<gtk4::TextView>() {
                        w.buffer().set_text(s);
                        true
                    } else {
                        err_out!("Incorrect widget type for text: {}", w.type_().name());
                        false
                    };
                }
            }

            err_out!("Data not a string");
            false
        }

        fn set_date_data(&self, w: &gtk4::Widget, value: &PropertyValue) -> bool {
            if let Some(d) = value.date() {
                return if let Some(w) = w.downcast_ref::<gtk4::Label>() {
                    w.set_text(&d.to_string());
                    true
                } else if let Some(w) = w.downcast_ref::<gtk4::Entry>() {
                    w.set_text(&d.to_string());
                    true
                } else {
                    err_out!("Incorrect widget type for date: {}", w.type_().name());
                    false
                };
            }

            err_out!("Data not a date");
            false
        }

        fn emit_metadata_changed(&self, prop: u32, value: &PropertyValue) {
            let mut props = MetadataPropertyBag::default();
            let mut old_props = MetadataPropertyBag::default();
            props.set_value(prop, value.clone());
            if let Some(old_val) = self
                .current_data
                .borrow()
                .as_ref()
                .and_then(|props| props.get(&prop))
            {
                old_props.set_value(prop, old_val.clone());
            }
            self.obj().emit_by_name::<()>(
                "metadata-changed",
                &[&WrappedPropertyBag(props), &WrappedPropertyBag(old_props)],
            );
        }
    }

    impl ObjectImpl for MetadataWidget {
        fn constructed(&self) {
            self.parent_constructed();

            self.widget.set_column_homogeneous(true);
            self.widget.set_row_homogeneous(false);
            self.widget.insert_column(0);
            self.widget.insert_column(0);
            self.widget.set_margin_start(8);
            self.widget.set_hexpand(true);
            self.obj()
                .upcast_ref::<super::ToolboxItem>()
                .set_child(Some(&self.widget));
        }

        fn signals() -> &'static [Signal] {
            use once_cell::sync::Lazy;
            static SIGNALS: Lazy<Vec<Signal>> = Lazy::new(|| {
                vec![
                    Signal::builder("metadata-changed")
                        .param_types([
                            <WrappedPropertyBag>::static_type(),
                            <WrappedPropertyBag>::static_type(),
                        ])
                        .run_last()
                        .build(),
                ]
            });
            SIGNALS.as_ref()
        }
    }

    #[glib::object_subclass]
    impl ObjectSubclass for MetadataWidget {
        const NAME: &'static str = "NpcMetadataWidget";
        type Type = super::MetadataWidget;
        type ParentType = super::ToolboxItem;

        fn new() -> Self {
            Self {
                widget: gtk4::Grid::new(),
                data_map: RefCell::new(HashMap::default()),
                current_data: RefCell::new(None),
                fmt: RefCell::new(None),
            }
        }
    }

    impl ToolboxItemImpl for MetadataWidget {}
    impl BoxImpl for MetadataWidget {}
    impl WidgetImpl for MetadataWidget {}
}
